File: examples/athletic-analyst-ts/src/mcp-bridge.ts
import { Client } from "@modelcontextprotocol/sdk/client/index.js";
import { StdioClientTransport } from "@modelcontextprotocol/sdk/client/stdio.js";
import type { Tool } from "@modelcontextprotocol/sdk/types.js";
import type { ExecutionTracer } from "./execution-trace.js";

const DATA_TOOL_NAMES = new Set(["get_health_metrics", "get_summary"]);
const AUTH_TOOL_NAMES = new Set(["direct_login", "oauth_login", "oauth_authorize"]);
const CONNECT_RETRIES = 3;
const CONNECT_RETRY_DELAY_MS = 2000;

export type ToolCallResult = {
  content: string;
  isError: boolean;
};

export type ExecutionSummary = {
  tool_calls: number;
  tool_failures: number;
  successful_data_calls: number;
  degraded: boolean;
  auth_ok: boolean;
};

export function buildMcpEnvironment(): Record<string, string> {
  const env = Object.fromEntries(
    Object.entries(process.env).filter((entry): entry is [string, string] => typeof entry[1] === "string")
  );

  const shouldForwardCustomBaseUrl =
    process.env.ATHLETIC_ANALYST_FORWARD_VYTALLINK_BASE_URL === "1";

  if (!shouldForwardCustomBaseUrl) {
    delete env.VYTALLINK_BASE_URL;
  }

  return env;
}

function isRetryableConnectError(error: unknown): boolean {
  const message = error instanceof Error ? error.message : String(error);
  return [
    "Backend server unavailable",
    "ECONNREFUSED",
    "fetch failed",
    "request to ",
  ].some((marker) => message.includes(marker));
}

function structuredSuccess(result: { structuredContent?: unknown }): boolean | undefined {
  const structured = result.structuredContent;
  if (typeof structured === "object" && structured !== null && "success" in structured) {
    return Boolean((structured as { success?: unknown }).success);
  }
  return undefined;
}

export class McpBridge {
  private client: Client;
  private tools: Tool[] = [];
  private executionSummary: ExecutionSummary = {
    tool_calls: 0,
    tool_failures: 0,
    successful_data_calls: 0,
    degraded: false,
    auth_ok: false,
  };

  constructor(private readonly tracer?: ExecutionTracer) {
    this.client = new Client({ name: "athletic-analyst", version: "1.0.0" });
  }

  async connect(): Promise<void> {
    let lastError: unknown;

    for (let attempt = 1; attempt <= CONNECT_RETRIES; attempt++) {
      this.client = new Client({ name: "athletic-analyst", version: "1.0.0" });
      this.tracer?.trace("bridge_connect_start", { attempt });

      const transport = new StdioClientTransport({
        command: "npx",
        args: ["@xmartlabs/vytallink-mcp-server"],
        env: buildMcpEnvironment(),
      });

      try {
        await this.client.connect(transport);
        const { tools } = await this.client.listTools();
        this.tools = tools;
        this.tracer?.trace("bridge_connect_end", { attempt, tools_loaded: tools.length });
        console.error(`[MCP] Connected — ${tools.length} tools loaded`);
        return;
      } catch (error) {
        lastError = error;
        const message = error instanceof Error ? error.message : String(error);
        this.tracer?.trace("bridge_connect_retry", { attempt, error: message });
        await this.client.close().catch(() => undefined);
        if (attempt >= CONNECT_RETRIES || !isRetryableConnectError(error)) {
          throw error;
        }
        console.error(`[MCP] Connection attempt ${attempt} failed: ${message}. Retrying...`);
        await new Promise((resolve) => setTimeout(resolve, CONNECT_RETRY_DELAY_MS));
      }
    }

    throw lastError instanceof Error ? lastError : new Error(String(lastError));
  }

  async close(): Promise<void> {
    await this.client.close();
  }

  getMcpTools(): Tool[] {
    return this.tools;
  }

  getExecutionSummary(): ExecutionSummary {
    return { ...this.executionSummary };
  }

  markDegraded(): void {
    this.executionSummary.degraded = true;
  }

  async callTool(name: string, args: Record<string, unknown>): Promise<ToolCallResult> {
    this.executionSummary.tool_calls += 1;
    const startedAt = Date.now();
    this.tracer?.trace("tool_start", { tool: name });

    try {
      const result = await this.client.callTool({ name, arguments: args });
      const content = result.content as Array<{ type: string; text?: string }>;
      const toolReportedError = Boolean(result.isError);
      const authSuccess = structuredSuccess(result as { structuredContent?: unknown });

      if (!content || content.length === 0) {
        return { content: "No data returned", isError: toolReportedError };
      }

      const renderedContent = content
        .map((block) => {
          if (block.type === "text") return block.text ?? "";
          return JSON.stringify(block);
        })
        .join("\n");

      if (toolReportedError) {
        this.executionSummary.tool_failures += 1;
        this.executionSummary.degraded = true;
      } else if (DATA_TOOL_NAMES.has(name)) {
        this.executionSummary.successful_data_calls += 1;
      }
      if (!toolReportedError && AUTH_TOOL_NAMES.has(name) && (authSuccess === undefined || authSuccess)) {
        this.executionSummary.auth_ok = true;
      }
      this.tracer?.trace("tool_end", {
        tool: name,
        duration_ms: Date.now() - startedAt,
        is_error: toolReportedError,
      });

      return { content: renderedContent, isError: toolReportedError };
    } catch (error) {
      const msg = error instanceof Error ? error.message : String(error);
      this.executionSummary.tool_failures += 1;
      this.executionSummary.degraded = true;
      this.tracer?.trace("tool_end", {
        tool: name,
        duration_ms: Date.now() - startedAt,
        is_error: true,
        error: msg,
      });
      console.error(`[MCP] Tool call failed: ${name} — ${msg}`);
      return { content: `Error calling tool ${name}: ${msg}`, isError: true };
    }
  }
}

File: examples/athletic-analyst-py/src/mcp_bridge.py
import asyncio
import contextlib
import json
import os
import sys
import time
from mcp import ClientSession, StdioServerParameters
from mcp.client.stdio import stdio_client

from .execution_trace import ExecutionTracer

DATA_TOOL_NAMES = {"get_health_metrics", "get_summary"}
AUTH_TOOL_NAMES = {"direct_login", "oauth_login", "oauth_authorize"}
CONNECT_RETRIES = 3
CONNECT_RETRY_DELAY_SECONDS = 2


def build_mcp_environment() -> dict[str, str]:
    env = dict(os.environ)
    should_forward_custom_base_url = (
        os.getenv("ATHLETIC_ANALYST_FORWARD_VYTALLINK_BASE_URL") == "1"
    )
    if not should_forward_custom_base_url:
        env.pop("VYTALLINK_BASE_URL", None)
    return env


def _is_retryable_connect_error(error: Exception) -> bool:
    message = str(error)
    return any(
        marker in message
        for marker in (
            "Backend server unavailable",
            "ECONNREFUSED",
            "fetch failed",
            "request to ",
        )
    )


def _structured_success(result) -> bool | None:
    structured = getattr(result, "structuredContent", None)
    if isinstance(structured, dict) and "success" in structured:
        return bool(structured["success"])
    return None


class McpBridge:
    def __init__(self, tracer: ExecutionTracer | None = None):
        self._session: ClientSession | None = None
        self._tools: list = []
        self._exit_stack = None
        self._tracer = tracer
        self._execution_summary = {
            "tool_calls": 0,
            "tool_failures": 0,
            "successful_data_calls": 0,
            "degraded": False,
            "auth_ok": False,
        }

    async def connect(self) -> None:
        last_error: Exception | None = None

        for attempt in range(1, CONNECT_RETRIES + 1):
            self._exit_stack = contextlib.AsyncExitStack()
            await self._exit_stack.__aenter__()
            self._session = None

            if self._tracer:
                self._tracer.trace("bridge_connect_start", attempt=attempt)

            try:
                server_params = StdioServerParameters(
                    command="npx",
                    args=["@xmartlabs/vytallink-mcp-server"],
                    env=build_mcp_environment(),
                )

                stdio_transport = await self._exit_stack.enter_async_context(
                    stdio_client(server_params)
                )
                read_stream, write_stream = stdio_transport
                self._session = await self._exit_stack.enter_async_context(
                    ClientSession(read_stream, write_stream)
                )

                await self._session.initialize()
                result = await self._session.list_tools()
                self._tools = result.tools
                if self._tracer:
                    self._tracer.trace(
                        "bridge_connect_end",
                        attempt=attempt,
                        tools_loaded=len(self._tools),
                    )
                print(f"[MCP] Connected — {len(self._tools)} tools loaded", file=sys.stderr, flush=True)
                return
            except Exception as error:
                last_error = error
                if self._tracer:
                    self._tracer.trace(
                        "bridge_connect_retry",
                        attempt=attempt,
                        error=str(error),
                    )
                await self.close()
                if attempt >= CONNECT_RETRIES or not _is_retryable_connect_error(error):
                    raise
                print(
                    f"[MCP] Connection attempt {attempt} failed: {error}. Retrying...",
                    file=sys.stderr,
                    flush=True,
                )
                await asyncio.sleep(CONNECT_RETRY_DELAY_SECONDS)

        if last_error is not None:
            raise last_error

    async def close(self) -> None:
        if self._exit_stack:
            await self._exit_stack.aclose()
            self._exit_stack = None
            self._session = None

    def get_mcp_tools(self) -> list:
        return self._tools

    def get_execution_summary(self) -> dict:
        return dict(self._execution_summary)

    def mark_degraded(self) -> None:
        self._execution_summary["degraded"] = True

    async def call_tool(self, name: str, args: dict) -> dict:
        self._execution_summary["tool_calls"] += 1
        if self._tracer:
            self._tracer.trace("tool_start", tool=name)
        started_at = time.perf_counter()
        if not self._session:
            self._execution_summary["tool_failures"] += 1
            self._execution_summary["degraded"] = True
            return {"content": "Error: MCP session not connected", "is_error": True}
        try:
            result = await self._session.call_tool(name, args)
            tool_reported_error = bool(getattr(result, "isError", False))
            if not result.content:
                if self._tracer:
                    self._tracer.trace(
                        "tool_end",
                        tool=name,
                        duration_ms=round((time.perf_counter() - started_at) * 1000),
                        is_error=tool_reported_error,
                    )
                if tool_reported_error:
                    self._execution_summary["tool_failures"] += 1
                    self._execution_summary["degraded"] = True
                return {"content": "No data returned", "is_error": tool_reported_error}
            parts = []
            for block in result.content:
                if hasattr(block, "text"):
                    parts.append(block.text)
                else:
                    parts.append(json.dumps(block.__dict__))
            rendered_content = "\n".join(parts)
            structured_success = _structured_success(result)
            if tool_reported_error:
                self._execution_summary["tool_failures"] += 1
                self._execution_summary["degraded"] = True
            elif name in DATA_TOOL_NAMES:
                self._execution_summary["successful_data_calls"] += 1
            if not tool_reported_error and name in AUTH_TOOL_NAMES and (
                structured_success is None or structured_success
            ):
                self._execution_summary["auth_ok"] = True
            if self._tracer:
                self._tracer.trace(
                    "tool_end",
                    tool=name,
                    duration_ms=round((time.perf_counter() - started_at) * 1000),
                    is_error=tool_reported_error,
                )
            return {"content": rendered_content, "is_error": tool_reported_error}
        except Exception as e:
            msg = str(e)
            self._execution_summary["tool_failures"] += 1
            self._execution_summary["degraded"] = True
            if self._tracer:
                self._tracer.trace(
                    "tool_end",
                    tool=name,
                    duration_ms=round((time.perf_counter() - started_at) * 1000),
                    is_error=True,
                    error=msg,
                )
            print(f"[MCP] Tool call failed: {name} — {msg}", file=sys.stderr, flush=True)
            return {"content": f"Error calling tool {name}: {msg}", "is_error": True}
